libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,507 @@
+ */
+
+#include <AP_IBus_Telem/AP_IBus_Telem.h>
+
+#if HAL_IBUS_TELEM_ENABLED
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_Arming/AP_Arming.h>
+#include <AP_GPS/AP_GPS.h>
+#include <AP_RPM/AP_RPM.h>
+
+void AP_IBus_Telem::init()
+{
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_IBUS_Telem, 0))) {
+        _port->set_options(_port->OPTION_HDPLEX);
+        hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_IBus_Telem::tick, void));
+    }
+}
+
+void AP_IBus_Telem::tick(void)
+{
+    ensure_listening();
+
+    uint16_t available = MIN(_port->available(), 1024U);
+    if (available == 0) {
+        return;
+    } else {
+        reset_read_state_after_timegap();
+    }
+
+    for (uint16_t i = 0; i < available; i++) {
+        const int16_t c = _port->read();
+        if (c < 0) {
+            return;
+        }
+
+        switch (_read_state) {
+        case ReadState::READ_LENGTH:
+            if (c != PROTOCOL_INCOMING_MESSAGE_LENGTH) {
+                _read_state = ReadState::DISCARD;
+                break;
+            }
+            _expected_checksum = 0xFFFF - c;
+            _read_state = ReadState::READ_COMMAND;
+            break;
+
+        case ReadState::READ_COMMAND:
+            _incoming_command = c & 0x0f0;
+            _incoming_sensor_id = c & 0x0f;
+            _expected_checksum -= c;
+            _read_state = ReadState::READ_CHECKSUM_HIGH_BYTE;
+            break;
+
+        case ReadState::READ_CHECKSUM_HIGH_BYTE:
+            _incoming_checksum_high_byte = c;
+            _read_state = ReadState::READ_CHECKSUM_LOW_BYTE;
+            break;
+
+        case ReadState::READ_CHECKSUM_LOW_BYTE:
+            if (_expected_checksum == (c << 8) + _incoming_checksum_high_byte) {
+                handle_incoming_message();
+            }
+            _read_state = ReadState::DISCARD;
+            break;
+
+        case ReadState::DISCARD:
+            break;
+        }
+    }
+}
+
+void AP_IBus_Telem::ensure_listening(void)
+{
+    if (!_initialized) {
+        _port->begin(115200);
+        _initialized = true;
+    }
+}
+
+void AP_IBus_Telem::reset_read_state_after_timegap(void)
+{
+    const uint32_t now = AP_HAL::millis();
+    if ((now - _last_received_message_time_ms) >= PROTOCOL_TIMEGAP) {
+        _read_state = ReadState::READ_LENGTH;
+        _last_received_message_time_ms = now;
+    }
+}
+
+void AP_IBus_Telem::handle_incoming_message()
+{
+    // Sensor 0 is reserved for the transmitter, so if for some reason it's requested we should ignore it
+    if (_incoming_sensor_id == 0) {
+        return;
+    }
+
+    // If we've reached the end of our sensor list, we shouldn't respond; this tells the receiver that there
+    // are no more sensors to discover.
+    if (
+        _incoming_sensor_id > sizeof(_sensors) / sizeof(_sensors[0])
```

Comment:
```suggestion
    if (_incoming_sensor_id > ARRAY_SIZE(_sensors) ||
```

Reply:
... actually, change this out for:
```
    const auto *sensor_defintiion = get_sensor_definition(_incoming_sensor_id);
    if (sensor_definition == nullptr || sensor_definition->sensor_type == 0) {
        return;
    }
```

... and change the calls to dependent functions to:
```
 handle_discover_command(*sensor_definition);
```

Reply:
With the change to `static const` that also removes the size ambiguity, meaning this check is now simply `_incoming_sensor_id > ARRAY_SIZE(_sensors)` with no parameter value checks.

However I've retained the `*sensor_definition` variable and passing it to the dependent functions.